### PR TITLE
[COST-5558] Fixing EC2 report to get the proper latest values

### PR DESCRIPTION
**Pull request by bacciotti**
## Jira Ticket

[COST-5558](https://issues.redhat.com/browse/COST-5558)

## Description

This change will create a new CTE and adapt ``reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql`` query to retrieve the proper latest values for the instances.

## Testing

*You can also test this change using the instructions on the [issue description](https://issues.redhat.com/browse/COST-5558) (**steps to reproduce**)*
1. Checkout Branch
2. Restart Koku
3. Ingest AWS data
4. It is recommended to compare the results with the previous version of the code (main branch) to ensure that the changes are working as expected.
    1. You should see no repeated values in ``reporting_awscostentrylineitem_summary_by_ec2_compute_p`` table
    2. You should see the latest values in ``reporting_awscostentrylineitem_summary_by_ec2_compute_p`` table

## Release Notes
- [ ] EC2 report now retrieves the proper latest values

```markdown
* [COST-5558](https://issues.redhat.com/browse/COST-5558) EC2 report now retrieves the proper latest values
```

### CODE REVIEW

File: koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql [hunk 1]
```diff
@@ -38,7 +38,29 @@ with cte_pg_enabled_keys as (
...
+    AND month = '{{month | sqlsafe}}'
+    AND lineitem_productcode = 'AmazonEC2'
+    AND product_productfamily LIKE '%Compute Instance%'
+    AND lineitem_resourceid != ''
+    AND lineitem_usagestartdate = (
+        SELECT max(date(lv.lineitem_usagestartdate)) as usage_start from hive.{{schema | sqlsafe}}.aws_line_items_daily as lv
+    )
+    GROUP BY
+        lineitem_resourceid,
+        product_instancetype
```
Comment (djnakabaale) [inline koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql:61]:
```suggestion
        lineitem_resourceid,
        product_instancetype,
        resourcetags,
        costcategory
```

File: koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql [hunk 2]
```diff
@@ -52,13 +74,13 @@ SELECT uuid() as uuid,
     memory,
     cast(
         map_filter(
-            cast(json_parse(tags) as map(varchar, varchar)),
+            cast(json_parse(tags[1]) as map(varchar, varchar)),
```
Comment (djnakabaale) [inline koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql:77]: I think we want all the tags 
```suggestion
            cast(json_parse(tags) as map(varchar, varchar)),
```

File: koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql [hunk 3]
```diff
@@ -52,13 +74,13 @@ SELECT uuid() as uuid,
...
             (k,v) -> contains(pek.keys, k)
         ) as json
-     ) as tags,
-    json_parse(costcategory) as cost_category,
+    ) as tags,
+    cast(json_parse(cost_category[1]) as json) as cost_category,
```
Comment (djnakabaale) [inline koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql:81]: Similar to tags, I think we all the cost category values
```suggestion
    cast(json_parse(cost_category) as json) as cost_category,
```

File: koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql [hunk 4]
```diff
@@ -79,82 +101,74 @@ SELECT uuid() as uuid,
...
+               OR lineitem_lineitemtype='Usage'
+               THEN lineitem_unblendedcost
+               ELSE savingsplan_savingsplaneffectivecost
+           END
+       ) as calculated_amortized_cost,
+       sum(pricing_publicondemandcost) as public_on_demand_cost,
+       max(pricing_publicondemandrate) as public_on_demand_rate
+    FROM hive.{{schema | sqlsafe}}.aws_line_items_daily as lid
+        JOIN cte_latest_values AS cte_l
+        ON lid.lineitem_resourceid = cte_l.resource_id
```
Comment (djnakabaale) [inline koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql:164]: Suggestion: 

I think this join should probably move up a level, maybe something similar to this;

```
CROSS JOIN cte_pg_enabled_keys AS pek
JOIN cte_latest_values AS cte_l ON ds.resource_id = cte_l.resource_id
LEFT JOIN postgres.{{schema | sqlsafe}}.reporting_awsaccountalias AS aa
```


and then use `cte_latest_values` results in the outer select statement

```
SELECT uuid() as uuid,
    -- ...
    resource_id,
    cte_l.instance_name,
    cte_l.instance_type,
    operating_system,
    region,
    vcpu,
    memory,
    cast(
        map_filter(
            cast(json_parse(cte_l.tags) as map(varchar, varchar)),
            (k,v) -> contains(pek.keys, k)
        ) as json
    ) as tags,
    cast(json_parse(cte_l.cost_category) as json) as cost_category,
```

File: koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql [hunk 5]
```diff
@@ -38,27 +38,54 @@ with cte_pg_enabled_keys as (
...
+        costcategory as cost_category
+    FROM hive.{{schema | sqlsafe}}.aws_line_items_daily as alid
+    WHERE source = '{{source_uuid | sqlsafe}}'
+    AND year = '{{year | sqlsafe}}'
+    AND month = '{{month | sqlsafe}}'
+    AND lineitem_productcode = 'AmazonEC2'
+    AND product_productfamily LIKE '%Compute Instance%'
+    AND lineitem_resourceid != ''
+    AND lineitem_usagestartdate = (
+      SELECT max(date(lv.lineitem_usagestartdate)) AS usage_start
```
Comment (myersCody) [inline koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql:57]: So, right now we only grab latest values for the latest day, but have you tested if a VM was spun down partially through the report period. Say that the VM was spun down on the 9th, would we still get the latest values for the 9th?

Comment (bacciotti) [reply koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql:57]: I think if we specify the source in the inner query, we can grab the latest from that specific VM.

Comment (myersCody) [reply koku/masu/database/trino_sql/reporting_awscostentrylineitem_summary_by_ec2_compute_p.sql:57]: I think the change you added to have the resource_ids match helped a ton here too:
```
trino:org1234567> SELECT
               ->     lineitem_resourceid as resource_id,
               ->     nullif(product_instancetype, '') as instance_type,
               ->     max(json_extract_scalar(json_parse(resourcetags), '$.Name')) AS instance_name,
               ->     resourcetags as tags,
               ->     costcategory as cost_category
               -> FROM aws_line_items_daily as alid
               -> WHERE source = 'b52d9062-c53d-4ae1-bdfe-606573aa7af6'
               -> AND year = '2024'
               -> AND month = '11'
               -> AND lineitem_productcode = 'AmazonEC2'
               -> AND product_productfamily LIKE '%Compute Instance%'
               -> AND lineitem_resourceid != ''
               -> AND lineitem_usagestartdate = (
               ->     SELECT max(date(lv.lineitem_usagestartdate)) AS usage_start
               ->     FROM aws_line_items_daily AS lv
               ->     WHERE lineitem_resourceid = alid.lineitem_resourceid
               ->     AND year = '2024'
               ->     AND month = '11'
               ->     AND source = 'b52d9062-c53d-4ae1-bdfe-606573aa7af6'
               -> )
               -> GROUP BY
               ->     lineitem_resourceid,
               ->     product_instancetype,
               ->     resourcetags,
               ->     costcategory;
 resource_id | instance_type |  instance_name  |                                                                                                                                                                                 tags               >
-------------+---------------+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------->
 i-444444442 | m5.large      | NULL            | {"CoM_RedHat_Rhel_varianT": "SAP", "com_REDHAT_rhel": "RHEL 7 ELS", "com_redhat_rhel_sla": "self-support", "com_redhat_rhel_usage": "disaster Recovery", "dashed-key-on-aws": "dashed-value", "envi>
 i-55555558  | m5.large      | NULL            | {"Map": "d2", "Mapping": "d1", "dashed-key-on-aws": "dashed-value", "version": "master"}                                                                                                           >
 i-444444445 | m5.large      | NULL            | {"CoM_RedHat_Rhel_varianT": "SAP", "com_REDHAT_rhel": "RHEL 8 ELS", "com_redhat_rhel_sla": "self-support", "com_redhat_rhel_usage": "disaster Recovery", "dashed-key-on-aws": "dashed-value", "envi>
 i-55555555  | m5.large      | instance-name-1 | {"Map": "a2", "Mapping": "a1", "Name": "instance-name-1", "dashed-key-on-aws": "dashed-value", "version": "prod"}                                                                                  >
 i-55555556  | m5.large      | instance-name-2 | {"CoM_RedHat_Rhel_varianT": "Workstation", "Map": "b2", "Mapping": "b1", "Name": "instance-name-2", "com_REDHAT_rhel": "8", "com_REDHAT_rhel_addon": "ELS", "com_redhat_rhel_conversion": "True", ">
 i-444444443 | m5.large      | instance-name-5 | {"CoM_RedHat_Rhel_varianT": "hPC", "Name": "instance-name-5", "com_REDHAT_rhel": "7", "com_redhat_rhel_sla": "self-support", "com_redhat_rhel_usage": "disaster Recovery", "dashed-key-on-aws": "da>
 i-444444446 | m5.large      | instance name 4 | {"CoM_RedHat_Rhel_varianT": "hPC", "Name": "instance name 4", "com_REDHAT_rhel": "8", "com_REDHAT_rhel_addon": "ELS", "com_REDHAT_rhel_conversion": "True", "com_redhat_rhel_sla": "self-support", >
 i-55555557  | m5.large      | instance_name_3 | {"CoM_RedHat_Rhel_varianT": "Workstation", "Map": "c2", "Mapping": "c1", "Name": "instance_name_3", "com_REDHAT_addon": "ELS", "com_REDHAT_conversion": "True", "com_REDHAT_rhel": "7", "com_redhat>
(8 rows)
```
Great Job Lucas!

**Comment by esebesto** [comment]
The PR was tested locally using IQE branch with updated tests

**Pull request merged**